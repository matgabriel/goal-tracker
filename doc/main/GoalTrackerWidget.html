<!DOCTYPE html><html lang="en"><head><title>main/GoalTrackerWidget</title></head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0"><meta name="groc-relative-root" content="../"><meta name="groc-document-path" content="main/GoalTrackerWidget"><meta name="groc-project-path" content="src/main/GoalTrackerWidget.js"><link rel="stylesheet" type="text/css" media="all" href="../assets/style.css"><script type="text/javascript" src="../assets/behavior.js"></script><body><div id="meta"><div class="file-path">src/main/GoalTrackerWidget.js</div></div><div id="document"><div class="segment"><div class="comments "><div class="wrapper"><h1 id="suivi-du-jour-pour-un-objectif">Suivi du jour pour un objectif</h1></div></div></div><div class="segment"><div class="code"><div class="wrapper"><span class="hljs-keyword">import</span> { hot } <span class="hljs-keyword">from</span> <span class="hljs-string">'react-hot-loader/root'</span>
<span class="hljs-keyword">import</span> React <span class="hljs-keyword">from</span> <span class="hljs-string">'react'</span>

<span class="hljs-keyword">import</span> Add <span class="hljs-keyword">from</span> <span class="hljs-string">'@material-ui/icons/Add'</span>
<span class="hljs-keyword">import</span> Fab <span class="hljs-keyword">from</span> <span class="hljs-string">'@material-ui/core/Fab'</span>
<span class="hljs-keyword">import</span> ThumbUp <span class="hljs-keyword">from</span> <span class="hljs-string">'@material-ui/icons/ThumbUp'</span>
<span class="hljs-keyword">import</span> Typography <span class="hljs-keyword">from</span> <span class="hljs-string">'@material-ui/core/Typography'</span>

<span class="hljs-keyword">import</span> { func, GoalPropType, nonNegativeInteger } <span class="hljs-keyword">from</span> <span class="hljs-string">'../shared/prop-types'</span>
<span class="hljs-keyword">import</span> Gauge <span class="hljs-keyword">from</span> <span class="hljs-string">'../shared/Gauge'</span></div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>Section de l&#39;écran principal, dédiée à un objectif.  Fournit notamment le
descriptif de l&#39;objectif et l’éventuel bouton de progression.</p></div></div><div class="code"><div class="wrapper"><span class="hljs-keyword">const</span> GoalTrackerWidget = ({</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>La déstructuration en force !</p></div></div><div class="code"><div class="wrapper">  goal,
  goal: { name, units, target },
  onProgress,
  progress,
}) =&gt; {</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>La beauté d&#39;un ternaire multi-lignes…</p></div></div><div class="code"><div class="wrapper">  <span class="hljs-keyword">const</span> adderComponent =
    target &gt; progress ? (
      <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-title">Fab</span>
        <span class="hljs-attribute">color</span>=<span class="hljs-value">'secondary'</span>
        <span class="hljs-attribute">onClick</span>=<span class="hljs-value">{()</span> =&gt;</span> onProgress &amp;&amp; onProgress(goal)}
        size='small'
      &gt;
        <span class="hljs-tag">&lt;<span class="hljs-title">Add</span> <span class="hljs-attribute">data-testid</span>=<span class="hljs-value">'in-progress'</span> /&gt;</span>
      <span class="hljs-tag">&lt;/<span class="hljs-title">Fab</span>&gt;</span>
    )</span> : (
      <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-title">Fab</span> <span class="hljs-attribute">disabled</span> <span class="hljs-attribute">size</span>=<span class="hljs-value">'small'</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-title">ThumbUp</span> <span class="hljs-attribute">data-testid</span>=<span class="hljs-value">'completed'</span> /&gt;</span>
      <span class="hljs-tag">&lt;/<span class="hljs-title">Fab</span>&gt;</span>
    )</span>

  <span class="hljs-keyword">return</span> (
    <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-title">div</span> <span class="hljs-attribute">className</span>=<span class="hljs-value">'goal'</span>&gt;</span>
      <span class="hljs-tag">&lt;<span class="hljs-title">div</span> <span class="hljs-attribute">className</span>=<span class="hljs-value">'summary'</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-title">Typography</span> <span class="hljs-attribute">variant</span>=<span class="hljs-value">'h6'</span> <span class="hljs-attribute">component</span>=<span class="hljs-value">'h2'</span>&gt;</span>
          {name}
        <span class="hljs-tag">&lt;/<span class="hljs-title">Typography</span>&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-title">Gauge</span> <span class="hljs-attribute">value</span>=<span class="hljs-value">{progress}</span> <span class="hljs-attribute">max</span>=<span class="hljs-value">{target}</span> /&gt;</span>
        <span class="hljs-tag">&lt;<span class="hljs-title">Typography</span> <span class="hljs-attribute">component</span>=<span class="hljs-value">'small'</span>&gt;</span>
          {`${progress} ${units} sur ${target}`}
        <span class="hljs-tag">&lt;/<span class="hljs-title">Typography</span>&gt;</span>
      <span class="hljs-tag">&lt;/<span class="hljs-title">div</span>&gt;</span>
      <span class="hljs-tag">&lt;<span class="hljs-title">div</span> <span class="hljs-attribute">className</span>=<span class="hljs-value">'cta'</span>&gt;</span>{adderComponent}<span class="hljs-tag">&lt;/<span class="hljs-title">div</span>&gt;</span>
    <span class="hljs-tag">&lt;/<span class="hljs-title">div</span>&gt;</span>
  )</span>
}</div></div></div><div class="segment"><div class="comments "><div class="wrapper"><p>Comme toujours, on définit les propriétés attendues/autorisées pour
validation.</p></div></div><div class="code"><div class="wrapper">GoalTrackerWidget.propTypes = {
  goal: GoalPropType.isRequired,
  progress: nonNegativeInteger.isRequired,
  onProgress: func,
}

<span class="hljs-keyword">export</span> <span class="hljs-keyword">default</span> hot(GoalTrackerWidget)</div></div></div></div></body></html>