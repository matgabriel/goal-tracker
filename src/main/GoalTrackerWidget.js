import { hot } from 'react-hot-loader/root'
import React from 'react'

import Add from '@material-ui/icons/Add'
import Fab from '@material-ui/core/Fab'
import ThumbUp from '@material-ui/icons/ThumbUp'
import Typography from '@material-ui/core/Typography'

import { func, GoalPropType, nonNegativeInteger } from '../shared/prop-types'
import Gauge from '../shared/Gauge'

const GoalTrackerWidget = ({
  goal,
  goal: { name, units, target },
  onProgress,
  progress,
}) => {
  const adderComponent =
    target > progress ? (
      <Fab
        color='secondary'
        onClick={() => onProgress && onProgress(goal)}
        size='small'
      >
        <Add data-testid='in-progress' />
      </Fab>
    ) : (
      <Fab disabled size='small'>
        <ThumbUp data-testid='completed' />
      </Fab>
    )

  return (
    <div className='goal'>
      <div className='summary'>
        <Typography variant='h6' component='h2'>
          {name}
        </Typography>
        <Gauge value={progress} max={target} />
        <Typography component='small'>
          {`${progress} ${units} sur ${target}`}
        </Typography>
      </div>
      <div className='cta'>{adderComponent}</div>
    </div>
  )
}

// Comme toujours, on définit les propriétés attendues/autorisées pour
// validation.
GoalTrackerWidget.propTypes = {
  goal: GoalPropType.isRequired,
  progress: nonNegativeInteger.isRequired,
  onProgress: func,
}

export default hot(GoalTrackerWidget)
