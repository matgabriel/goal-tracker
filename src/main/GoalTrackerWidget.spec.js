import React from 'react'
import { cleanup, fireEvent, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import GoalTrackerWidget from './GoalTrackerWidget'

describe('<GoalTrackerWidget />', () => {
  const goal = {
    id: '0123456789abcdef01234567',
    name: 'My goal',
    target: 42,
    units: 'wombats',
  }

  describe('when not completed', () => {
    it('should render appropriately', () => {
      for (const progress of [0, 1, 21, 41]) {
        const { getByRole, queryByTestId, queryByText } = render(
          <GoalTrackerWidget goal={goal} progress={progress} />
        )
        expect(getByRole('heading')).toHaveTextContent(goal.name)
        expect(getByRole('progressbar')).toHaveAttribute(
          'aria-valuenow',
          String(Math.round((progress * 100) / goal.target))
        )
        expect(
          queryByText(`${progress} ${goal.units} sur ${goal.target}`)
        ).not.toBeNull()

        expect(queryByTestId('in-progress')).not.toBeNull()

        cleanup()
      }
    })

    it('should trigger its onProgress on click', () => {
      const progress = 21
      const onProgress = jest.fn()
      const { getByRole } = render(
        <GoalTrackerWidget
          goal={goal}
          onProgress={onProgress}
          progress={progress}
        />
      )

      fireEvent.click(getByRole('button'))

      expect(onProgress).toHaveBeenCalledTimes(1)
      expect(onProgress).toHaveBeenCalledWith(goal)
    })

    it('should otherwise match the expected snapshot', () => {
      const { container } = render(
        <GoalTrackerWidget goal={goal} progress={21} />
      )

      expect(container).toMatchSnapshot()
    })
  })

  describe('when completed (or exceeded)', () => {
    it('should render appropriately', () => {
      for (const progress of [goal.target, goal.target + 1, goal.target + 10]) {
        const { queryByTestId } = render(
          <GoalTrackerWidget goal={goal} progress={progress} />
        )

        expect(queryByTestId('in-progress')).toBeNull()
        expect(queryByTestId('completed')).not.toBeNull()

        cleanup()
      }
    })

    it('should otherwise match the expected snapshot', () => {
      const { container } = render(
        <GoalTrackerWidget goal={goal} progress={42} />
      )

      expect(container).toMatchSnapshot()
    })
  })
})
