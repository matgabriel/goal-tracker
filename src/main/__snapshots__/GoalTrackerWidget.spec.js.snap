// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GoalTrackerWidget /> when completed (or exceeded) should otherwise match the expected snapshot 1`] = `
<div>
  <div
    class="goal"
  >
    <div
      class="summary"
    >
      <h2
        class="MuiTypography-root MuiTypography-h6"
      >
        My goal
      </h2>
      <div
        aria-valuenow="100"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate"
        role="progressbar"
        style="height: 8px;"
      >
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
          style="transform: translateX(0%);"
        />
      </div>
      <small
        class="MuiTypography-root MuiTypography-body1"
      >
        42 wombats sur 42
      </small>
    </div>
    <div
      class="cta"
    >
      <button
        class="MuiButtonBase-root MuiFab-root MuiFab-sizeSmall Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            data-testid="completed"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M1 21h4V9H1v12zm22-11c0-1.1-.9-2-2-2h-6.31l.95-4.57.03-.32c0-.41-.17-.79-.44-1.06L14.17 1 7.59 7.59C7.22 7.95 7 8.45 7 9v10c0 1.1.9 2 2 2h9c.83 0 1.54-.5 1.84-1.22l3.02-7.05c.09-.23.14-.47.14-.73v-2z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<GoalTrackerWidget /> when not completed should otherwise match the expected snapshot 1`] = `
<div>
  <div
    class="goal"
  >
    <div
      class="summary"
    >
      <h2
        class="MuiTypography-root MuiTypography-h6"
      >
        My goal
      </h2>
      <div
        aria-valuenow="50"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate"
        role="progressbar"
        style="height: 8px;"
      >
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
          style="transform: translateX(-50%);"
        />
      </div>
      <small
        class="MuiTypography-root MuiTypography-body1"
      >
        21 wombats sur 42
      </small>
    </div>
    <div
      class="cta"
    >
      <button
        class="MuiButtonBase-root MuiFab-root MuiFab-sizeSmall MuiFab-secondary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiFab-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            data-testid="in-progress"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
